server/aesdsocket.c

```diff
@@ -158,146 +137,351 @@ int main(int argc, char* argv[])
+ */
+void SignalHandler(int signo)
+{
+	syslog(LOG_INFO, "Caught signal, exiting");
+	terminate = true;
+
+	// Close all threads
+	struct threadInfo_s* ptr = NULL;
+	while(!SLIST_EMPTY(&head))
+	{
+		ptr = SLIST_FIRST(&head);
+		if(ptr->connectionInProgress == false)
+		{
+			pthread_join(ptr->threadId, NULL);
+			SLIST_REMOVE_HEAD(&head, entries);
+			free(ptr);
+		}
+	}
+
+	int rc = remove(AESD_SOCKET_DATA_FILE);
+	if(rc == -1)
+		perror("remove socket data file");
+	if(serverSockfd != -1)
+		close(serverSockfd);
+
+	exit(0);
+
+}
+
+/*
+ * Function to send file content through socket
+ */
+void SendFileDataToClient(int fileFd, int socketFd)
+{
+	ssize_t ret;
+	char buf[512];
+
+	// Seek to start of file
+	lseek(fileFd, 0, SEEK_SET);
+
+	while ((ret = read (fileFd, buf, sizeof buf)) != 0)
+	{
+        if (ret == -1) 
+		{
+                if (errno == EINTR)
+                    continue;
+                perror ("read");
+                break;
+        }
+
+		// Send whatever is read to client
+		send(socketFd, buf, (ret/sizeof(char)), 0);
+	}
+}
+
+int SetTimer()
+{
+	struct itimerval delay;
+
+	delay.it_value.tv_sec = 10;
+    delay.it_value.tv_usec = 0;
+    delay.it_interval.tv_sec = 10;
+    delay.it_interval.tv_usec = 0;
+
+
+	int rc = setitimer(ITIMER_REAL, &delay, NULL);
+	if(rc)
+	{
+		perror("setitimer");
+		return -1;
+	}
+	return 0;
+}
+
+int ClearTimer()
+{
+	struct itimerval delay;
+
+	delay.it_value.tv_sec = 0;
+    delay.it_value.tv_usec = 0;
+    delay.it_interval.tv_sec = 0;
+    delay.it_interval.tv_usec = 0;
+
+
+	int rc = setitimer(ITIMER_REAL, &delay, NULL);
+	if(rc)
+	{
+		perror("cleartimer");
+		return -1;
+	}
+	return 0;
+}
+
+void WriteTimestampToFile(int fd)
+{
+	char buf[100];
+	memset(buf, 0, 100);
+
+	time_t t = time(NULL);
+  	struct tm currentTime = *localtime(&t);
+
+	snprintf(buf, 100, "timestamp: %d-%02d-%02d %02d:%02d:%02d\n", currentTime.tm_year + 1900, currentTime.tm_mon + 1, currentTime.tm_mday, currentTime.tm_hour, currentTime.tm_min, currentTime.tm_sec);
+	write(fd, buf, strlen(buf) * sizeof(char));
+}
+
+void CleanUpList()
+{
+	// Clean up the list
+	struct threadInfo_s* ptr = NULL;
+	struct threadInfo_s* nextPtr = NULL;
+	SLIST_FOREACH_SAFE(ptr, &head, entries, nextPtr)
+	{
+		if(ptr->connectionInProgress == false)
+		{
+			// This thread can be removed
+			pthread_join(ptr->threadId, NULL);
+			totalConnections--;
+			SLIST_REMOVE(&head, ptr, threadInfo_s, entries);
+			free(ptr);
+		}
+	}
+}
+
+/*
+ * Function for handling alarm signal
+ */
+void AlarmHandler(int signo)
+{
+	hasTimeElapsed = true;
+	CleanUpList();
+	if(totalConnections == 0)
+	{
+		int fd = fd = open(AESD_SOCKET_DATA_FILE, O_RDWR | O_APPEND | O_CREAT, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
 		if(fd == -1)
 		{
 			perror("socket data file open");
-			close(acceptedSocketfd);
-			freeaddrinfo(serverInfo);
-			close(serverSockfd);
-			return -1;
+			exit(-1);
 		}
+		WriteTimestampToFile(fd);
```

Comment:
@suryakanteti see assignment requirements about locking.  Are all of these operations safe to do from a signal handler?